WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Tags/ProductTagsViewController.swift

```diff
@@ -0,0 +1,558 @@
+            textView.selectedRange = NSRange(location: range.location - 1, length: 0)
+            textViewDidChange(textView)
+            return false
+        } else if range.length == 0, // Inserting
+            text == ",", // a comma
+            range.location == original.length // at the end
+        {
+            // Append a space
+            textView.text = original.replacingCharacters(in: range, with: ", ")
+            textViewDidChange(textView)
+            return false
+        } else if text == "\n", // return
+            range.location == original.length, // at the end
+            !partialTag.isEmpty // with some (partial) tag typed
+        {
+            textView.text = original.replacingCharacters(in: range, with: ", ")
+            textViewDidChange(textView)
+            return false
+        } else if text == "\n" // return anywhere else
+            {
+                return false
+        }
+        return true
+    }
+
+    private func normalizeText() {
+        // Remove any space before a comma, and allow one space at most after.
+        let regexp = try! NSRegularExpression(pattern: "\\s*(,(\\s|(\\s(?=\\s)))?)\\s*", options: [])
+        let text = textView.text ?? ""
+        let range = NSRange(location: 0, length: (text as NSString).length)
+        textView.text = regexp.stringByReplacingMatches(in: text, options: [], range: range, withTemplate: "$1")
+    }
+
+    /// Normalize tags for initial set up.
+    ///
+    /// The algorithm here differs slightly as we don't want to interpret the last tag as a partial one.
+    ///
+    private func normalizeInitialTags(tags: [String]) -> String {
+        var tags = tags.filter({ !$0.isEmpty })
+        tags.append("")
+        return tags.joined(separator: ", ")
+    }
+
+    private func updateSuggestions() {
+        dataSource.selectedTags = completeTags
+        dataSource.searchQuery = partialTag
+        tableView.reloadData()
+    }
+}
+
+extension ProductTagsViewController: UITableViewDelegate {
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        tableView.deselectRow(at: indexPath, animated: true)
+        switch tableView.dataSource {
+        case is FailureDataSource:
+            loadTags()
+        case is LoadingDataSource:
+            return
+        case is SuggestionsDataSource:
+            suggestionTapped(cell: tableView.cellForRow(at: indexPath))
+        default:
+            assertionFailure("Unexpected data source")
+        }
+    }
+
+    private func suggestionTapped(cell: UITableViewCell?) {
+        guard let tag = cell?.textLabel?.text else {
+            return
+        }
+        complete(tag: tag)
+    }
+}
+
+// MARK: - Data Sources
+
+private protocol ProductTagsDataSource: UITableViewDataSource {
+    var selectedTags: [String] { get set }
+    var searchQuery: String { get set }
```

Comment:
It was done in that way because the datasource used in `ProductTagsViewController` is declared as a generic `ProductTagsDataSource`, and when it's a `SuggestionsDataSource` we need to update the `selectedTags` and the `searchQuery`, but I agree with you that in this case, they are superfluous. I did a change here 2d89dd9918922dace3e1f554a9295e8d5cf75838. Can you take a second look at it? ☺️